network/vpack/vpack.go

```diff
@@ -0,0 +1,384 @@
+// Copyright (C) 2019-2025 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package vpack
+
+import (
+	"fmt"
+	"math/bits"
+)
+
+// A vote is made up of 14 values, some of which are optional.
+// The required values are: cred.pf, r.rnd, r.snd, sig.p, sig.p1s, sig.p2,
+// sig.p2s, sig.s (sig.ps is always zero).
+// The remaining 6 optional values are either present or omitted, and their
+// presence is indicated in a 1-byte bitmask in the header.
+const (
+	bitPer    uint8 = 1 << iota // r.per
+	bitDig                      // r.prop.dig
+	bitEncDig                   // r.prop.encdig
+	bitOper                     // r.prop.oper
+	bitOprop                    // r.prop.oprop
+	bitStep                     // r.step
+
+	propFieldsMask      uint8 = bitDig | bitEncDig | bitOper | bitOprop
+	totalRequiredFields       = 8
+)
+
+const headerSize = 2 // 1 byte for mask, 1 byte for future use
+
+// StatelessEncoder compresses a msgpack-encoded vote by stripping all msgpack
+// formatting and field names, replacing them with a bitmask indicating which
+// fields are present. It is not thread-safe.
+type StatelessEncoder struct {
+	cur  []byte
+	pos  int
+	mask uint8
+
+	requiredFields uint8
+}
+
+// NewStatelessEncoder returns a new StatelessEncoder.
+func NewStatelessEncoder() *StatelessEncoder {
+	return &StatelessEncoder{}
+}
+
+// CompressVoteBound estimates the maximum size needed for compressed vote data
+func CompressVoteBound(srcSize int) int {
+	// Vote messages typically get ~80% smaller, but we will use
```

Comment:
I guess it can't possibly matter, but you _know_ it will get smaller by at least removing all the msgpack names, right?